File: src/modules/discussion/service.py
"""토론 스레드 및 댓글 생성/조회 서비스."""
import uuid
from datetime import datetime, timezone
from typing import Optional

from modules.discussion.comment import DiscussionComment
from modules.discussion.repository import DiscussionRepository
from modules.discussion.thread import DiscussionThread


class DiscussionService:
    """
    토론 스레드 및 댓글 생성과 조회를 담당하는 서비스.

    스레드/댓글 생성 시 id와 생성 시각을 부여하고 저장소에 위임한다.
    스레드 상태 전환, 로그 기록 등의 세부 동작은 이후 태스크에서 채워지므로,
    이 서비스는 생성과 조회만 제공하는 골격이다.
    """

    def __init__(self, repository: DiscussionRepository):
        """
        서비스를 초기화한다.

        Args:
            repository: 토론 저장소
        """
        self.repository = repository

    async def create_thread(
        self, document_id: str, title: str, created_by: str
    ) -> DiscussionThread:
        """
        새로운 토론 스레드를 생성한다.

        Args:
            document_id: 스레드가 속할 문서의 id
            title: 스레드 제목
            created_by: 스레드를 생성하는 사용자의 id

        Returns:
            생성된 토론 스레드
        """
        thread = DiscussionThread(
            id=str(uuid.uuid4()),
            document_id=document_id,
            title=title,
            created_by=created_by,
            created_at=datetime.now(timezone.utc),
        )
        return await self.repository.create_thread(thread)

    async def get_thread(self, id: str) -> Optional[DiscussionThread]:
        """
        주어진 id로 토론 스레드를 조회한다.

        Args:
            id: 조회할 토론 스레드의 고유 식별자

        Returns:
            조회된 토론 스레드 또는 없으면 None
        """
        return await self.repository.get_thread(id)

    async def list_threads_by_document_id(
        self, document_id: str
    ) -> list[DiscussionThread]:
        """
        주어진 문서의 토론 스레드를 생성 순서대로 나열한다.

        Args:
            document_id: 조회할 문서의 고유 식별자

        Returns:
            문서의 토론 스레드 목록 (생성 순서)
        """
        return await self.repository.list_threads_by_document_id(document_id)

    async def add_comment(
        self, thread_id: str, body: str, created_by: str
    ) -> DiscussionComment:
        """
        토론 스레드에 새로운 댓글을 추가한다.

        Args:
            thread_id: 댓글이 속할 스레드의 id
            body: 댓글 본문
            created_by: 댓글을 작성하는 사용자의 id

        Returns:
            생성된 댓글
        """
        comment = DiscussionComment(
            id=str(uuid.uuid4()),
            thread_id=thread_id,
            body=body,
            created_by=created_by,
            created_at=datetime.now(timezone.utc),
        )
        return await self.repository.create_comment(comment)

    async def list_comments_by_thread_id(
        self, thread_id: str
    ) -> list[DiscussionComment]:
        """
        주어진 스레드의 댓글을 생성 순서대로 나열한다.

        Args:
            thread_id: 조회할 스레드의 고유 식별자

        Returns:
            스레드의 댓글 목록 (생성 순서)
        """
        return await self.repository.list_comments_by_thread_id(thread_id)

File: php/src/Modules/Discussion/Service.php
<?php

declare(strict_types=1);

namespace MintWiki\Discussion;

/**
 * 토론 스레드/댓글 생성·조회 서비스 (태스크 0711).
 *
 * Python `DiscussionService`(src/modules/discussion/service.py)의
 * `create_thread`/`get_thread`/`list_threads_by_document_id`/`add_comment`/
 * `list_comments_by_thread_id` 다섯 흐름만 옮긴다. close/reopen/pause_thread,
 * hide_comment, audit_recorder 연동은 스레드 상태 전이·댓글 모더레이션이라
 * 이 태스크(저장/조회 왕복) 범위 밖이다.
 *
 * 빈 본문/작성자 검증은 새로 만들지 않는다 — `Thread`/`Comment` 생성자가
 * 이미 `Empty*Error`를 던지므로, 이 서비스는 id/생성 시각만 채워 생성자에
 * 위임한다. 스레드 생성 시 초기 상태는 `Thread`의 기본값 문자열 대신
 * `ThreadState::Open`을 명시적으로 전달해, 두 타입이 같은 값을 쓴다는
 * 사실을 코드로도 드러낸다 — `Thread`가 `ThreadState`를 검증에 쓰지
 * 않는다는 기존 설계(Thread.php docblock)는 그대로 유지된다.
 *
 * 클래스 이름은 `docs/php-namespace-mapping.md`가 고정한 규칙대로 이미
 * `MintWiki\Discussion` namespace 안에 있으므로 중복되는 `Discussion`
 * 접두어를 뺀다 (Document\Service와 동일한 패턴).
 */
final class Service
{
    public function __construct(private readonly Repository $repository)
    {
    }

    /**
     * 새로운 토론 스레드를 생성한다.
     *
     * @throws EmptyThreadDocumentIdError documentId가 비어있거나 공백만 있는 경우
     * @throws EmptyThreadTitleError title이 비어있거나 공백만 있는 경우
     * @throws EmptyThreadCreatedByError createdBy가 비어있거나 공백만 있는 경우
     */
    public function createThread(string $documentId, string $title, string $createdBy): Thread
    {
        $thread = new Thread(
            self::generateId(),
            $documentId,
            $title,
            $createdBy,
            self::nowUtc(),
            ThreadState::Open->value
        );

        return $this->repository->createThread($thread);
    }

    /**
     * 주어진 id로 토론 스레드를 조회한다. 없으면 null을 반환한다.
     */
    public function getThread(string $id): ?Thread
    {
        return $this->repository->getThread($id);
    }

    /**
     * 주어진 문서의 토론 스레드를 생성 순서대로 나열한다.
     *
     * @return Thread[]
     */
    public function listThreadsByDocumentId(string $documentId): array
    {
        return $this->repository->listThreadsByDocumentId($documentId);
    }

    /**
     * 토론 스레드에 새로운 댓글을 추가한다.
     *
     * @throws EmptyCommentThreadIdError threadId가 비어있거나 공백만 있는 경우
     * @throws EmptyCommentBodyError body가 비어있거나 공백만 있는 경우
     * @throws EmptyCommentCreatedByError createdBy가 비어있거나 공백만 있는 경우
     */
    public function addComment(string $threadId, string $body, string $createdBy): Comment
    {
        $comment = new Comment(self::generateId(), $threadId, $body, $createdBy, self::nowUtc());

        return $this->repository->createComment($comment);
    }

    /**
     * 주어진 스레드의 댓글을 생성 순서대로 나열한다.
     *
     * @return Comment[]
     */
    public function listCommentsByThreadId(string $threadId): array
    {
        return $this->repository->listCommentsByThreadId($threadId);
    }

    private static function nowUtc(): \DateTimeImmutable
    {
        return new \DateTimeImmutable('now', new \DateTimeZone('UTC'));
    }

    /**
     * UUID v4 문자열을 생성한다 (스레드/댓글 id 발급용, `Document\Service`와
     * 동일한 방식).
     */
    private static function generateId(): string
    {
        $bytes = random_bytes(16);
        $bytes[6] = chr((ord($bytes[6]) & 0x0f) | 0x40);
        $bytes[8] = chr((ord($bytes[8]) & 0x3f) | 0x80);
        $hex = bin2hex($bytes);

        return sprintf(
            '%s-%s-%s-%s-%s',
            substr($hex, 0, 8),
            substr($hex, 8, 4),
            substr($hex, 12, 4),
            substr($hex, 16, 4),
            substr($hex, 20, 12)
        );
    }
}
